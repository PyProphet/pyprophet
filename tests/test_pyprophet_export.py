from __future__ import print_function

import os
import subprocess
import shutil
import sys

import pandas as pd
import sqlite3

import pytest

pd.options.display.expand_frame_repr = False
pd.options.display.precision = 4
pd.options.display.max_columns = None

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")

def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += str(subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT))
    except subprocess.CalledProcessError as error:
        print(error, end="", file=sys.stderr)
        raise
    return stdout


def _run_osw(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # peptide-level
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=global"

    # protein-level
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=global"

    # export
    cmdline += " export --in=test_data.osw --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    if not peptide:
        cmdline += " --no-peptide"

    if not protein:
        cmdline += " --no-protein"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_osw_unscored(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)

    # export
    cmdline = "pyprophet export --in=test_data.osw --out=test_data.tsv --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_ipf(regtest, temp_folder, transition_quantification=False, ipf="disable"):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms1 --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # MS2-level
    cmdline += " score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # transition-level
    cmdline += " score --in=test_data.osw --level=transition --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # IPF
    cmdline += " ipf --in=test_data.osw"

    # export
    cmdline += " export --in=test_data.osw --no-peptide --no-protein --ipf_max_peptidoform_pep=1  --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    cmdline += " --ipf=" + ipf

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_unscored(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # export
    cmdline = "pyprophet export-compound --in=test_data_compound.osw --out=test_data_compound_unscored.tsv --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_unscored.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_compound_ms1(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS1-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms1 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms1.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms1.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_ms2(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS2-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms2 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms2.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms2.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)


def test_osw_0(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, False)

def test_osw_1(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, True, False, False)

def test_osw_2(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, True, False)

def test_osw_3(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, True)

def test_ipf_0(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "disable")

def test_ipf_1(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, True, "disable")

def test_ipf_2(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "peptidoform")

def test_ipf_3(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "augmented")

def test_compound_0(tmpdir, regtest):
    _run_compound_ms1(regtest, tmpdir.strpath)
    
def test_compound_1(tmpdir, regtest):
    _run_compound_ms2(regtest, tmpdir.strpath)

def test_compound_unscored(tmpdir, regtest):
	_run_compound_unscored(regtest, tmpdir.strpath)
	
def test_osw_unscored(tmpdir, regtest):
	_run_osw_unscored(regtest, tmpdir.strpath)
