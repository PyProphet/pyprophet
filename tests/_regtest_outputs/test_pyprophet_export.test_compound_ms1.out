   Adducts  Charge  Intensity       RT  aggr_prec_Peak_Apex  aggr_prec_Peak_Area  assay_RT  assay_rt      compound_name  d_score  decoy  delta_RT  delta_rt                                           filename                   id  leftWidth  m_score        mz  peak_group_rank  rightWidth               run_id   sum_formula  transition_group_id
0    M+Na+       1  71301.300  475.327              90356.0           2.8526e+05   475.718  475.7176     Isoprothiolane   4.7037      0    -0.391   -0.3906  /Users/alka/Documents/work/projects/OpenSWATH_...  6732968656885689606    472.409   0.0081  313.0539                1     478.992  2408145804652532658    C12H18O4S2                    3
1    M+Na+       1  19160.700  475.303              31671.0           9.9949e+04   475.718  475.7184     Isoprothiolane   3.5837      0    -0.415   -0.4154  /Users/alka/Documents/work/projects/OpenSWATH_...  6921299947704833207    472.408   0.0231  313.0539                1     479.814  1007350642398073598    C12H18O4S2                    3
2    M+Na+       1   9308.450  475.253              17956.0           5.0582e+04   475.718  475.7180     Isoprothiolane   3.9994      0    -0.465   -0.4650  /Users/alka/Documents/work/projects/OpenSWATH_...  5388675400702073898    472.408   0.0200  313.0539                1     479.814   214379053049545951    C12H18O4S2                    3
3    M+Na+       1    147.925  481.026                746.0           6.9487e+02   475.718  475.7176     Isoprothiolane   2.9940      0     5.308    5.3084  /Users/alka/Documents/work/projects/OpenSWATH_...  7508128245913369262    479.815   0.0442  313.0539                2     483.106  2408145804652532658    C12H18O4S2                    3
4     M+H+       1    949.648  453.263            1705988.0           4.6946e+06   453.946  453.9459  Triadimefon_decoy   3.4105      1    -0.683   -0.6829  /Users/alka/Documents/work/projects/OpenSWATH_...   671489720159068151    451.012   0.0266  294.1004                1     461.710  3140099155998074833  C14H16ClN3O2                    6
..     ...     ...        ...      ...                  ...                  ...       ...       ...                ...      ...    ...       ...       ...                                                ...                  ...        ...      ...       ...              ...         ...                  ...           ...                  ...
95    M+H+       1   1788.530  455.453               5767.0           1.6677e+04   455.082  455.0823           Boscalid   4.3533      0     0.371    0.3707  /Users/alka/Documents/work/projects/OpenSWATH_...  4911838227523744331    452.659   0.0110  343.0399                1     460.888  6344784885616328495  C18H12Cl2N2O                  104
96   M+Na+       1  57212.200  502.251              97318.0           3.1908e+05   501.656  501.6563      Picoxystrobin   3.1736      0     0.595    0.5947  /Users/alka/Documents/work/projects/OpenSWATH_...   995120417245308422    499.563   0.0319  390.0924                1     506.969  3140099155998074833   C18H16F3NO4                  108
97   M+Na+       1  28159.500  502.291              32319.0           8.4656e+04   501.656  501.6562      Picoxystrobin   4.2847      0     0.635    0.6348  /Users/alka/Documents/work/projects/OpenSWATH_...  6098343142262325573    499.564   0.0110  390.0924                1     505.324  2408145804652532658   C18H16F3NO4                  108
98   M+Na+       1  11019.300  502.517               7266.0           2.0389e+04   501.656  501.6556      Picoxystrobin   3.0171      0     0.861    0.8614  /Users/alka/Documents/work/projects/OpenSWATH_...  3727023492293302868    499.564   0.0430  390.0924                1     506.147  1007350642398073598   C18H16F3NO4                  108
99   M+Na+       1   5384.930  502.411               4580.0           1.1208e+04   501.656  501.6560      Picoxystrobin   3.3918      0     0.755    0.7550  /Users/alka/Documents/work/projects/OpenSWATH_...  7379182434544455482    499.564   0.0281  390.0924                1     505.324   214379053049545951   C18H16F3NO4                  108

[100 rows x 23 columns]
